### ISSUE TITLE
multicore access on Raspberry Pi Pico

### ISSUE THREAD

**Issue by mlewus** [adafruit/circuitpython#4106]
The Pi Pico has 2 physical cores, but only one core is usable in CircuitPython. Micropython has limited multicore functionality when used with the pico, allowing the user to start a separate task while passing variables in the task call, like this:

```
def mytask(pin, delay):
# bang away at a pin or whatever, call with
import _thread
_thread.start_new_thread(mytask, (GP2, 0.2))
```
mytask runs independently of the main mcu core, and runs until it returns or mcu reset. 

Is this planned for inclusion in Circuitpython?

**Comment by tannewt** [comment]
It isn't a short-term goal. We have many other APIs to implement before we do multi-core.

I'm considering an API to facilitate running native code on the second core instead of additional Python. What are you using it to do?

**Comment by daveythacher** [comment]
I understand in theory how to load the code. I understand in theory what you can do with that second core that may be of interest. In theory I understand the appeal of Python here and how it could be terrible yet passable.

What I do not understand is how you plan to implement it and provide something like real time. I am also curious to know if you know where that issue could exist.

The RP2040 provides the raw materials to achieve it. They are cost optimized and will not play nice with malloc. Planning is an aspect I also wonder about. How many people will understand the pitfalls here? 

This is a tone switch for Python. Python has a few roles by itself.

**Comment by daveythacher** [comment]
malloc may be better than I thought initially. They strip the heap, which creates a pipeline of sort. Would be interesting to know latency. I am not sure this changes everything, but makes the common case better. That is neat.

**Comment by mlewus** [comment]
Not sure how relevant this is to CircuitPython, but I did some multicore testing with MicroPython on a Pico back in February. I found that global variables require about 1mSec to get updated between cores. The only element that updates faster is a lock. These must be interrupt driven as they update very fast, on the order of a few microseconds.

**Comment by Red-M** [comment]
I'd really like this for a project I'm building to allow the 2nd core to handle a bunch of additional IO tasks while the first core is on input detection.

**Comment by mlewus** [comment]
My limited understanding is that Circuitpython is, or started out as, a fork of Micropython. Micropython supports multitasking, locks, and inter-task globals on the Pico. Can a Circuitpython developer comment as to the difficulty of implementing this on Circuitpython?

**Comment by daveythacher** [comment]
> I'd really like this for a project I'm building to allow the 2nd core to handle a bunch of additional IO tasks while the first core is on input detection.

This would be a valid usage for this. Use one core for IO, DSP, computation, interface, real time, etc. The memory design of the RP2040 is fairly strong. However the IO design is not as strong. DMA and CPU can compete for the lower and upper IO region. However you could partition the upper region to one and the other to the other one.

Note I have not looked at this significantly. DMA generally does compete at some point. However blocking the stall condition on IO bus is common with secondary data bus for DMA. I have not looked that far into it.

IPC is also solved with the RP2040's design. This would likely be the default method of communication between these two. However certain memory partitioning could be interesting for this secondary core. Core 0 owns everything and circuitPython makes prolific usage of dynamic memory. 

Meaning one could impose on the others memory usage. If you reserve a decent static memory allocation for core 1 then you will be removing memory from circuitPython. This could create confusion. However if you decide to force core 1 into dynamic memory you could mess up the logic of core 1.

Again I believe the raw materials for this exist. However how exactly circuitPython decides to play it will be interesting. CircuitPython has kind of worked itself into a corner with this Java-like approach.

**Comment by daveythacher** [comment]
> My limited understanding is that Circuitpython is, or started out as, a fork of Micropython. Micropython supports multitasking, locks, and inter-task globals on the Pico. Can a Circuitpython developer comment as to the difficulty of implementing this on Circuitpython?

MicroPython uses python for both cores. I believe the suggestion here is to use C/C++ or assembly for the secondary core. Each has there own set of confusion. Python on both may be a little more straightforward in high level manner.

**Comment by tannewt** [comment]
> My limited understanding is that Circuitpython is, or started out as, a fork of Micropython. Micropython supports multitasking, locks, and inter-task globals on the Pico. Can a Circuitpython developer comment as to the difficulty of implementing this on Circuitpython?

It's really hard to say because none of us have looked into it.

Right now we're based on MicroPython ~1.9.4 (https://github.com/micropython/micropython/commit/25ae98f07cb3c4488cb955403dfe56b8bb8db6f0) which is a year or two old. I think upstream has made some concurrency improvements since then so the first step I'd take is to merge in newer MicroPython. There is an outstanding PR to do that here: #4280 

Once that is done, I'd try enabling the settings in CircuitPython to see if they work. You'd also need to look at the MP changes for the RP2040's second core. That'll likely expose many areas where there are concurrency issues that need to be fixed.

**Comment by Red-M** [comment]
Personally, I'd prefer if CircuitPython offered "more difficult" to use features such as `_thread` because I really just want to get this done and I know how to handle myself with concurrency and memory access.

I'd really want Python on the 2nd core because frankly all my IO tasks are laid out with CircuitPython libraries, so if I had to write C for it I might instead just move to MicroPython instead and re-implement the libraries I'm currently using in MicroPython.

**Comment by mlewus** [comment]
The problem for me is that using circuitpython on the Pico means throwing away half of the machine. I don't really see C++ on the 2nd core as a solution. Most people who use circuitpython do so for the advantages that it offers over C++. I'm not sure how the decision making works as to which features get to the top of the list but it would be great to see this get prioritized. I've been using micropython, but that means rewriting a lot of the hardware drivers that already exist in Circuitpython. So that is not ideal.

**Comment by Red-M** [comment]
Would this then be better suited to being tracked along/after 7.x has a release so that there is more control around concurrency?

**Comment by ladyada** [comment]
@Red-M if its *just* for keymatrix management, it would probably be better to adapt _gamepad to be a matrix scanning helper...? then it isnt multi-core dependant and could run on other chips :)

**Comment by Red-M** [comment]
@ladyada Thanks for the reply, but I've gotten the keymatrix poll/scan down to just interrupt pin polling which is way faster.
I'm making a full sized keyboard with RGB and other features so I'm really wanting access to the other core with all the IO libraries.

**Comment by ladyada** [comment]
if the keymatrix was handled in the background - what other IO would you need?

**Comment by Red-M** [comment]
dotstar and displayio.

**Comment by ladyada** [comment]
kk - displayio is also handled in the background with DMA and dotstars are near instantaneous - you can drive them at 8mhz. im not convinced you'd gain anything with dual core support and its not going to be adding to circuitpython soon

**Comment by mlewus** [comment]
Frankly I think that's a little short-sighted. But it's your football. Closing this issue.

**Comment by ladyada** [comment]
reopening because its a valid issue. 
for some purposes - a keyboard, for example - we really do think it does not need a dual core. we are trying to help get projects done and not have to wait for something that does not exist!

**Comment by Red-M** [comment]
I have to agree with @mlewus.
You can't just wait for another cycle to get more feedback to handle other tasks, _near_ instant isn't nearly as fast as real time.

**Comment by ladyada** [comment]
in python the garbage collector can run at anytime - theres no such thing as true guaranteed real time task handling in interpretted python which is why you'd have to do it C. 

i'd like to find a solution to the problem, however the problem isnt clear yet - USB tasks out every 1ms, its not 'instant' either. some really specific numbers or complete example code would be useful. if it cannot be measured, we cannot improve it :)

**Comment by dhalbert** [comment]
The [`adafruit-blinka`](https://github.com/adafruit/Adafruit_Blinka) Python library, which is mostly used for Linux, emulates many of the native modules provided by CircuitPython. But besides supporting Linux, it also has a MicroPython wrapper. So you can still use CircuitPython libraries, on top of MicroPython, if you do indeed want to take advantage of something only MicroPython provides.

**Comment by dhalbert** [comment]
Some keyboard projects done in CircuitPython:
https://github.com/KMKfw/kmk_firmware
https://github.com/tannewt/ckd63 (4 years old)

If you'd like to discuss this in more detail, this issue is probably not place. [Discord](https://adafru.it/discord) or the [forums](https://forums.adafruit.com/viewforum.php?f=60) are good.

**Comment by mlewus** [comment]
@dhalbert, I'll take a look at adafruit-blinka and continue this in a more appropriate forum. Thanks for your help!

**Comment by daveythacher** [comment]
@ladyada This can be done, however circuitPython will need to scope it in some manner. C/C++ API is one approach, which manages resources. This may be required even for the Python version, but hidden. There are other approaches. How you plan to scope the role of the secondary core is probably the leading question. The official scope of circuitPython is indirectly referenced here.

There is a coherency issues regardless in at least three places. This is kind of true for C and Python. However Python technically can ignore some aspects given the RP2040's implementation. However there is a consistency problem for all of circuitPython, which this could break. Your call.

@Red-M You could use two circuitPython instances via two boards. Lot more expensive, but would solve the problem in the mean time. This gets you parallel execution and modularization. However uses different programming model, which is not an issue here technically.

**Comment by Red-M** [comment]
@dhalbert Just letting you know and others in case they see this issue but that suggestion doesn't work due to adafruit/Adafruit_Python_PlatformDetect#144.
Anyway, I think I'll be moving off adafruit libs all together.

**Comment by ladyada** [comment]
yeah you cannot use gamepad (yet) because it doesnt support an expander. `blinka` is great but will not add hid support to micropython-on-pico. if you need help getting your project optimized you can come by discord

**Comment by tannewt** [comment]
> The problem for me is that using circuitpython on the Pico means throwing away half of the machine. I don't really see C++ on the 2nd core as a solution. Most people who use circuitpython do so for the advantages that it offers over C++. I'm not sure how the decision making works as to which features get to the top of the list but it would be great to see this get prioritized. I've been using micropython, but that means rewriting a lot of the hardware drivers that already exist in Circuitpython. So that is not ideal.

The RP2040 isn't our only platform and the second core isn't the only thing we can't do that the hardware supports. It's all a question of prioritization and Adafruit-funded folks are prioritizing other work. (We discuss this in places like this issue and the weekly meeting.) We'd love to have threading and multicore support but it's just not clear to us that the benefit is worth time cost. I always prioritize PRs so no one outside of Adafruit would be blocked on implementing this.

I'd encourage you to try CircuitPython for your specific project. There may be other ways to optimize the code so that you can do what you want from CircuitPython even with just one core. (I know @Red-M got their keyboard code going by optimizing I2C transactions when scanning.)

**Comment by daveythacher** [comment]
Valid usage: Crypto-Processor. May support 3Mbps AES-256 with 133MHz. M3/M4 may do 13Mbps with 80MHz.

https://www.bearssl.org/speed.html
https://github.com/Ko-/aes-armcortexm

Valid usage: Graphics Processor. Drawing Mandelbrot algorithm can be stressful on lower end CPUs.

Valid usage: IO Controller. Tight timing and event resolution.

Valid usage: Signal Processor. AI/ML/DSP/etc.

Singleton for secondary core? Make supported classes implement runnable? Not sure I am crazy about this but it would scope it. Providing a means to hide details in implementation. (Avoid some conflicts at runtime?)

**Comment by daveythacher** [comment]
Another interesting implementation of dual core controller is dsPIC33CH. This may be supportable under microPython, but will not likely be supported under circuitPython. The secondary core there would easily support native due to its design. The controllers are completely independent aside from mailbox, FIFO, PPS, etc. The secondary could would not likely support Python.

**Comment by mlewus** [comment]
@tannewt I completely understand that Adafruit has limited resources and has to make the best use of them to support business & community goals. I think Adafruit is a great example of a well-run tech company. It's the reason that I buy products from you whenever I can.

**Comment by daveythacher** [comment]
Here is some of fun of native code:
https://github.com/micropython/micropython/issues/ 583

Referenced here:
https://github.com/adafruit/circuitpython/issues/ 2999
https://github.com/adafruit/circuitpython/issues/ 2825

**Comment by Red-M** [comment]
I understand that there is limited time and resource for concurrency but more and more devices/MCUs are supporting/providing more than 1 core.

It would be great to have this in general and not just for the RP2040 also for various reasons/usecases. I'd honestly prefer some planning go into getting concurrency as a whole (if it isn't/hasn't been done yet) because more devices in the future will have a multicore offering.

Offering even just async tasks to be placed onto other cores that share the same IO would be excellent, as much as I'd prefer to do threading workloads.

**Comment by mlewus** [comment]
@Red-M, a week ago I would have agreed with you. But one of the Adafruit developers brought up a valid point. Which is, they support 100s of boards with circuitpython. Only a few of those are multi-core. Given the vastly greater scope of requested functionality with respect to available development labor, I understand why this has not been done.

**Comment by Red-M** [comment]
@mlewus I'm not saying it isn't going to be a lot of work but some form of concurrency planning should be done soon as more MCUs are going to come out with more than 1 core.

**Comment by sirkha** [comment]
Adding to the application conversation, I have a Feather RP2040 with a Pimoroni Envior+ wing, an AMG8833 wing, an adalogger wing, and a PoE wing all attached to a single trippler that also has an IR tranceiver on it. I would like to be able to use the 2nd core to process the analog audio from the mic on the enviro+ and possibly stream it over the network to do voice command recognition while first core handles the rest of the sensors and U/I.

**Comment by mlewus** [comment]
@Red-M, yes of course you're right. I think we're going to see more low end multi-core microcontrollers. And there are some tasks that are too complex for a Pio but for which you really don't want to be interrupted.

**Comment by john--** [comment]
> @Red-M, yes of course you're right. I think we're going to see more low end multi-core microcontrollers. And there are some tasks that are too complex for a Pio but for which you really don't want to be interrupted.

Agreed. I'm happy to see multi-core support is being considered!

**Comment by nmorse** [comment]
up-vote for native code on the second processor, I have a project that requires 4 rotary-encoders and presents a HID interface.

- in my experience, reading encoders is helped by a little physics (momentum) modeling. A bit too complicated for PIO, but C code is working great on the second processor. 
- for HID handling, circuitpython shines there, contrasted with getting HID to work via the C-sdk, not easy.

Possibly this is an example of an app that needs dual-cores: circuitpython on Proc0, native (driver) code on Proc1?

**Comment by ladyada** [comment]
@nmorse did you try circuitpython as is? 4 rotary encoders and HID is not processor intensive.

**Comment by nmorse** [comment]
I will try that, (on this project) up till now I had been doing all 'C-sdk', thanks

**Comment by razzededge** [comment]
I also would vote for this - even some kind of port/wrapper for _threads from MicroPython would be nice.

**Comment by mlewus** [comment]
My FWIW if this is to be prioritized: We need to show demonstrable instances where a real project, preferably with publicly posted code, that can't be made to work with the current circuitpython, but could be done with multiple core support. Or, ones that could be made to perform meaningfully better with multiple core support. 

So far there have been a lot of what if kind of requests here, including one of mine, but none that meet that standard. Adafruit has limited resources and many demands on them. If we want those resources applied to this feature, we need to describe a better use case than had been shown so far. Or, live without it until either one of us takes it on ourselves or more processors come out with multiple cores, which would provide more use cases for this code.

**Comment by Red-M** [comment]
We've given them examples and they've been discarded, just because something isn't compute intensive doesn't mean it doesn't need multithread access or couldn't benefit from it. Without interrupts its pretty hard to make something with tight timings and just because someone thinks that isn't possible just because of a GC still doesn't mean it can't be worked around and still perform better.

**Comment by ladyada** [comment]
hiya folks, lets keep comments positive so they cant be misread - we're here to solve issues, sometimes we can solve issues without waiting for the circuitpython team to do a lot of work.

we recently added `asyncio`, working with micropython team so that our codebases and libraries could be compatible. there is work happening with adding more advanced features: circuitpython has a lot of async stuff going on and as we gain more dual 
core processors, there'll be more work on it.

we want to make sure that what we write is well-supported and usable. adding use cases will help us greatly to make sure we have coverage - we think we captured the most common async/irq needs so far and there will be more!

**Comment by Red-M** [comment]
I think if you want to meet coverage of end-user usecases, you might want to look at how Python handles that in their unit tests.

**Comment by rmilby13** [comment]
@ladyada Use Case: I would like to use one core for monitoring a Loconet (from Digitrax) communication channel and a DCC reception (using PIO for the low level work) and use the other core for managing 16 GPIO (servos/buttons/current block detection/IR sensors, maybe other uses later), updating a neopixel bus, i2c communication (RFID, eeprom, display, etc) and possibly a bit more. I want to keep the loconet and DCC on it's own core to help prevent data loss in the PIO I/O buffers. I'm just getting started now, so I haven't tried to load everything on one core yet.. It may handle it, but as the RP2040 has dual cores I would like to leverage one just for communications and one for "local work".

**Comment by bluelasers** [comment]
@rmilby13 That is a good idea. However I would recommend you look into using two RP2040s. Do you want to create a repo with discussions enabled where we could chat about it? If you want to use a single RP2040 this gets more complicated. There is a lot to talk about and they do not like that stuff here. I honestly am not sure they will get around to this, since its the only multicore MCU they support. Any meaningful API for building this would need their support, so that is too bad. They could make some money on this, but I doubt they want the baggage.

**Comment by tomasinouk** [comment]
Hi, I came here looking for a solution to a problem, where RP2040 communicates with LTE/WiFi to send data. At this time pretty much any library is using `time.sleep`. Meaning while I am sending data over UART the code is blocking. The device does not react to buttons, etc. My understanding is if a `thread` would be used for offloading data another `thread` could be used to keep the device "operational". I am happy to be educated on different approach to resolve this scenario. 
Thank you

**Comment by AdaRoseCannon** [comment]
The use case I have is very superficial but I want to run animations on the macropad screen, whilst having the LEDs fade smoothly. Unfortunately if I am writing big changes to the screen the rate the LEDs update slows down and it looks a little off. 

So I want to run my displayio things in the other thread so that the standard keyboard features can run indepdently.

**Comment by agrabbs** [comment]
Possible use case: we have a sensor logger application that reads from a 6dof and writes to the onboard flash of the RP2040. We poll the sensor as many times as possible in a span of 1 to 10 seconds and have tried various methods of saving the data:

1. One sensor poll, one IO write (or append): Slowest, highest amount of sensor data loss > ~50%
2. Poll sensor into array and write at end (anything over 2 seconds and the RP2040 runs out of ram): Fastest, but we cannot run for more than 2-3 seconds and the memory is full. 
3. Poll sensors into array and write to file at X interval: Fast, but we lose about ~150 sensor data points during each IO write. 

We would love to see a library that allowed utilization of the second core so we can offload the IO write to another core while the sensor data collection continues.

**Comment by ladyada** [comment]
@agrabbs you are not CPU bound, you are flash-memory-speed bound - writing to any flash memory requires erasing blocks, which will take a long time AND the same internal flash memory is being read from for the RP2040 to run. 

this is very hard to do. ideally you can buffer into an array - you should have ~ 100k available. or, FRAM is the easiest external-memory solution as it has instantaneous writes. ESP32-Sx with PSRAM might be reasonable because you get SRAM speed but 2MB of space

**Comment by RyanDam** [comment]
I'm building a small step sequencer + synthesizer using raspberry pico. This project involves keypad+encoder, screen and audio output. Currently, everything is running in just 1 core and sometimes the audio glitch because of heavy UI rendering. It's will be best if I can utilize the second core for just audio processing while the first core is dedicated to handling user input and UI rendering. By utilizing the second core, maybe more advanced audio processing can be archivable.

**Comment by bradanlane** [comment]
My usecase is mostly an up ode of @AdaRoseCannon example:

Display update in one thread while “everything else” is in another thread.

_(I know of a micropython project where they went through macerations to get this working but it made world of difference in the end result.)_

**Comment by bradanlane** [comment]
I have another usecase _which may be niche_ …

charlieplexing required an infinite loop to keep the LEDs updated. It would be convenient to run this on a second core. It would require some form of shared memory or communications to choose which LEDs need to be active.

**Comment by jepler** [comment]
It may be possible to do charlieplexing with features now in CircuitPython, without multicore. It's possible to repeatedly send a buffer to a PIO peripheral forever with [`background_write`](https://docs.circuitpython.org/en/latest/shared-bindings/rp2pio/index.html#rp2pio.StateMachine.background_write), which opens a ton of interesting possibilities.

There's an example on the Learn system to use this to drive a [4-digit 7-segment LED directly](https://learn.adafruit.com/intro-to-rp2040-pio-with-circuitpython/advanced-driving-seven-segment-displays-with-background-writes).

In an example I coded but didn't demonstrate in the guide, I combined the LED segment driving code with the advanced PWM code and controlled the brightness of segments individually: https://github.com/adafruit/Adafruit_CircuitPython_PIOASM/blob/main/examples/pioasm_7seg_fader.py

I think the same structure would adapt readily to charlieplexed LEDs.

**Comment by deshipu** [comment]
The `_pew` module does LED matrix updating without either a second core or pio, as it does it on samd21. The matrix happens to not be charlieplexed, but I have successfully used variants of the same code with a charlieplexed matrix as well.

**Comment by Rybec** [comment]
My project is also audio synthesis.  Timing is critical, and I'm _trying_ to find a way to do it with the PIO, but the FIFO is extremely small for buffering audio, and with CircuitPython I don't have the option of triggering an interrupt to run the generation code as needed.  I haven't written any code yet, so I'm not sure exactly how limited I'm going to be yet, but the ideal situation would be to be able to use the second core to generate the audio and the first to handle user input and the display.

Interestingly, while I _was_ looking for way to just run Python code on the second core, native C would be even better for this particular project, because timing would be far easier to manage, and speed wouldn't be as much of an issue.  (Also, I have some basic audio synth software written in C already...)  Long term, I would love to see both Python and C/C++ as options for this.  The ability to run native code directly from CircuitPython on the RP2040 would open up a _ton_ of options, but not every multi-threaded project needs that.

But yeah, I would really like to write an audio synth using the RP2040, and it would be ideal if it was a "second core" module that just uses the second core so that the first is free for whatever else.  And I'm not sure how feasible this is with just the one core.

I do have some things I can try.  PyRTOS might be able to manage scheduling in a way that isn't too heavy handed.  (I added service routines a while back that might be ideal for this...)  I can use a simpler timing loop algorithm.  I might be able to use asyncio, but it is designed for I/O bound tasks, and audio synthesis is CPU bound.  I don't think any of these solutions would allow the synthesizer code to be released as a solid standalone synth module though, since they would be locked into a serial timing algorithm that any user would have to work within.  (And I really would like to turn the synth portion specifically into an open source project, but I don't want to force the user to adhere to a restrictive scheduling regimen.)


Anyhow, I figured if you are looking for applications where this kind of feature would be critical, I've got one!  I can always go with MicroPython instead, but if I'm going to write a full on synth module, I would really rather do it for CircuitPython.

**Comment by tannewt** [comment]
We've got #7218 going to discuss a coprocessor API that would apply to running native code on the second core (not Python on the second core.)

@Rybec you may be interested in [`synthio`](https://docs.circuitpython.org/en/latest/shared-bindings/synthio/index.html). Though it needs to be changed to take in real-time midi messages too.

**Comment by Rybec** [comment]
Thanks for the suggestion (I noticed you've worked on audio for CircuitPython a bit, so I thought you might have something to add!), but I'm not trying to do MIDI stuff, and it looks like `synthio` currently can only do a square wave.  I'm looking for something more full featured, where I can basically programmatically craft custom waveforms in real-time.  Or rather, I _can_ programmatically craft waveforms already.  I just need an API that allows me to play them in real-time, as they are being generated.  (I'll definitely go check out the other issue for the coprocessor API thing!)

It looks like I might be able to manage with `audiocore`, `audiopwmio`, and `audiomixer`.  `audiomixer` provides the ability to easily loop multiple voices and control them independently.  This allows me to generate a single wave period as an `audiocore.RawSample`, play it on loop, and avoid having to constantly generate audio in real-time, which will keep processor time down.  The downside is that without a FIFO style buffer, I can't reliably do lead-in and fade-out effects on instruments (as looping the waveform would constantly repeat those, and I don't have sufficiently precise timing with `audiomixer`/CircuitPython to transition from a lead-in to the main waveform to a fade-out smoothly).  Also, generating waves within a limited number of samples means that where the frequency is not divisible by the sample rate, the pitch is subject to rounding.  I'm not too worried about that right now though. Yesterday, I was able to successfully generate and play square, sine, sawtooth, and triangle waves (the four foundational waveforms used in audio synthesis) this way, and the generation lag isn't detectable so far.  The next step is creating composite waveforms from those to change the timbre, and I suspect the RP2040 will handle that reasonably well as long as I don't get too complex.

Know what would be really cool though?  Something similar to `audiocore.RawSample`, but with a FIFO instead of a static array.  If the FIFO had some way of easily checking how many additional samples it has room for, it would be really easy to include a function as part of the program loop that checks the FIFO and generates enough to refill it each loop, before it runs out.  Even a 50 to 100 millisecond FIFO buffer would probably be plenty.  (Maybe this should be a feature request specifically for `audiocore`?)

**Comment by tannewt** [comment]
> Know what would be really cool though? Something similar to `audiocore.RawSample`, but with a FIFO instead of a static array. If the FIFO had some way of easily checking how many additional samples it has room for, it would be really easy to include a function as part of the program loop that checks the FIFO and generates enough to refill it each loop, before it runs out. Even a 50 to 100 millisecond FIFO buffer would probably be plenty. (Maybe this should be a feature request specifically for `audiocore`?)

That does sound neat! I'm totally open to merging in a new module to do that. Unfortunately, you clearly know way more about it than I do. So, I'm not the best person to do it.

**Comment by Rybec** [comment]
I'll see if I can take a look at the `audiocore` code.  I don't think I have enough knowledge of the RP2040, let alone all of the other microcontrollers that might be able to support this, to code it myself.  I do know that the RP2040 has some kind of DMA thing that can be used to drive a PWM at a specified sample rate.  If 'audiocore` is already using that, I might be able to work out how to set that up with a FIFO for the RP2040.  I don't have a ton of free time right now though, so I don't think I would be able to do it for the entire CircuitPython ecosystem.  Maybe if I can pull it off for the RP2040, others can use that as a starting point to expand it?  I can't make any promises at this point, but I'll try!

**Comment by tannewt** [comment]
> Maybe if I can pull it off for the RP2040, others can use that as a starting point to expand it?

Totally fine to start with one port.

I'd suggest starting with RawSample. The DMA to PWM is already done for RP2040 since CircuitPython has its own rudimentary audio pipeline. Try changing RawSample to queue input samples instead of returning the same buffer over and over again.

**Comment by Rybec** [comment]
This is what I was thinking.  Thank you for verifying that it's already using DMA to PWM.  Knowing that really helps.  I have some questions about how the API calls to RawSample from the audio pipeline work, but I'll save those for later.  I think I'm going to study the code a bit more and then open an issue to ask questions and such.

**Comment by tannewt** [comment]
#circuitpython-dev on [Discord](https://adafru.it/discord) is the best place to ask dev questions.

**Comment by Rybec** [comment]
I thought you might say that.  I've joined, and I'll ask there instead then.  Thanks for the help thus far!

**Comment by mobigroup** [comment]
There is one more use case for IR receiver. That’s possible to detect IR codes on RPI Pico using CircuitPython but that’s impossible to do something more, see for example https://learn.adafruit.com/ir-sensor/circuitpython for IR receiver only (CircuitPython code) and https://learn.adafruit.com/remote-controlled-led-candelabra/code for IR receiver plus LEDs (C code). Even block code language for MicroBit allows to use IR receiver and neopixels together via interrupts but CircuitPython doesn’t. Do we need to have one RPI pico for an IR receiver and one more for neopixels and so on? Maybe is that much better to have interrupts and threads in CircuitPython instead? Actually, I’m confused by the programming language which requires a separate computer for any task like to single sensor or a LED processing.

**Comment by chrisalbertson** [comment]
I was Googling to find out how to run a task on the second CPU and found this is still unimplemented in Circuitpython.   I'm shocked.  Having dual cores is one of the "top three" features of the RP2 and the reason for the "2" in "RP2".    Excuse the pun but this is a core feature of the chip.   

What is the use case?  
task 1 - reading data from a bunch of sensors and we can't stop looking
task 2 - interacting with another computer via serial link and running commands that come over the link

I have a project running on ESP32/Micropython that has five tasks.   It works really well and is so easy to understand.  Shared global data and locks work as expected.

Yes I know how to write my own cooperative multitasking in python but it is not really parallel unless there are multiple cores.

**Comment by KenSamson** [comment]
As an evolution of a use case, this is probably similar to other audio requests...

I will be honest in that this may be ambitious for Python on an an S3, though my hunch now is that it could work.   That S3 hums along....

The idea is to take in audio, encode it into a 
smaller byte stream, and connect to a gateway on the internet.  It's ham radio over internet in this case.

The internet side already exists with hundreds of thousands of users.  ( android example included...)

Being able to process audio from a microphone in one core, 
and send M17 encoded audio in IP over the other core, would be a beautifil thing.

Then we also have to run the gui, etc.

This is an android app showing a running voice app...
This is a superset, I would not have all the same settings and functions....
https://youtu.be/Jqm_1cowo2g

The project that already has c code running for the voice encoding:
https://m17project.org/

https://github.com/M17-Project

Would that have to get added as C or worked in python, I dont know yet.  

https://www.kb6nu.com/m17-an-open-source-dmr-like-system/

Something with built-in microphones could be ideal...

https://www.adafruit.com/product/5290?gclid=Cj0KCQiAtvSdBhD0ARIsAPf8oNlAaQAcU95kfvMU_4_zs4ZGsFmO7cFvVtrmZdrHMY81zaUIIysFAiwaAmHGEALw_wcB

I have had this rumbling around my head, and at least thought I would share.

**Comment by chrisalbertson** [comment]
You definatly do not need two cores to stream audio over the Internet.   It can be done using cooperative multitasking.  The Python asyncio package on one core would work, assuming the one core is fast enough.   With audio streaming you trade latency for glitches.  Each end has a buffer of audio samples queued up so if a task is blocked for less than the length of the buffer the sound is still OK.   

User interface don't have to block, they are usually callback based.   You do need a fast core to do all this, but there is no logical need to have two cores.

**Comment by mobigroup** [comment]
@chrisalbertson We are not able to do realtime audio processing and perform internet connection at the same time on a single core because fixed resources required to record and encode audio and fill the transmission buffer and that’s impossible to add unpredictably long network operations on the same core. DMA allows to output the prepared buffer but for network transmission we need to have a second core.

**Comment by KenSamson** [comment]
@chrisalbertson  When we talk streaming, perhaps enough optimization might get there,  though it's not trivial.  When you add encoding which can be viewed as the use of computing resources to optimize and reduce the size if the stream to fit through a narrow wireless modem, that makes it harder.  That is what M17 is all about.

**Comment by chrisalbertson** [comment]
I did not mean dual core is not helpful, just that logically it is not required if the single core is fast enough.


I have an application running now where I use both core of the RP2040.  The same code also runs on the dual core ESP32.  I am evaluating which is best.

When I found that circuitpython did not allow use of the second core, I switched to micropython.

If transcoding on the RP2040 is slow, try an ESP32.   The ES32 is a little fastr and it has hardware floating point so it is much beter at numeric computation.  It also has built-in WiFi so the streamed audio couple be placed on-line by the ESP32 itself.

Also look into Ulab port of numpy.   If you can write your algorithm in terms of a numpy array then it will be much fastr in numpy.

Or course you can always write the transcoding function in C.

**Comment by steffenrog** [comment]
Current project ongoing, also requires multicores.

I want to listen to CAN (Adafruit MCP2515 lib), for instructions for LED settings.
While also reading analog, and digital inputs and send those over can.

Obviously the adafruit libs are very handy, but the no access to both cores makes it "not very usefull"

**Comment by AdrieK** [comment]
> There is one more use case for IR receiver. That’s possible to detect IR codes on RPI Pico using CircuitPython but that’s impossible to do something more, see for example https://learn.adafruit.com/ir-sensor/circuitpython for IR receiver only (CircuitPython code) and https://learn.adafruit.com/remote-controlled-led-candelabra/code for IR receiver plus LEDs (C code). Even block code language for MicroBit allows to use IR receiver and neopixels together via interrupts but CircuitPython doesn’t. Do we need to have one RPI pico for an IR receiver and one more for neopixels and so on? Maybe is that much better to have interrupts and threads in CircuitPython instead? Actually, I’m confused by the programming language which requires a separate computer for any task like to single sensor or a LED processing.

I really do appreciate the hard work that people do on circuitpython and I really am impressed what has been achieved in the recent years.
But, having said that, having no interrupts is a shame (polling feels so '80s), no threading is a pity to say the least and being not able to benefit from multiple cores is making me real sad - having half of my chips idling through the day. I recently did some tests comparing my old-fashioned Atmega328P (16 MHz in an Arduino Uno) with an Itsybitsy M4 (120MHz on circuitpython) and other circuitpython boards and I was surprised about the speed difference (in favour of the ancient 328P)
ESP32, running freertos on both cores is really a joy to use in this perspective.

In my humble opinion the community would really benefit from microcontroller features that are (becoming) standard. 

Let me repeat: I do appreciate all the hard work, and I do understand (something of) the challenges ahead so I don't want to put the blame on anyone.

**Comment by chrisalbertson** [comment]
On Wed, Feb 22, 2023 at 6:13 AM Adrie ***@***.***> wrote:

> In my humble opinion the community would really benefit from
> microcontroller features that are (becoming) standard.
>
>
Then move on to Micropython.   It is quite a lot ahead of Circuitpython.
There is no down side to switching.

> Let me repeat: I do appreciate all the hard work
>
The work would be much more appreciated if it were not wasted on this
Circuitpython fork, but rather put into advancing Micropython.  We'd all be
further along

-- 

Chris Albertson
Redondo Beach, California

**Comment by SonalPinto** [comment]
I guess anyone who is looking for more juice out of an RP2040 running CircuitPython ends up here...

Possible usecase (and perhaps builds upon [AdaRoseCannon's comment](https://github.com/adafruit/circuitpython/issues/4106#issuecomment-1137754472)) - A keyboard with the main core running [KMK](https://github.com/KMKfw/kmk_firmware) (keyboard firmware managing the matrix scan and HID services) and displayio animation/management running on the second core to produce fancy procedurally generated animations like this [glitch effect](https://github.com/SonalPinto/mt-choc/blob/main/docs/firmware.md) ([Mt Choc keyboard](https://github.com/SonalPinto/mt-choc)). The animation compute has been trimmed down to the bare minimum with some tilegrid tricks, but otherwise could have been unchained if it had a core to itself.

**Comment by Rybec** [comment]
I guess I'll give an update on something I did last year (or maybe earlier this year) where this is relevant:

I wrote a CircuitPython (in C) to allow for playback of real-time generated audio, for use in synths.  (No, I haven't pushed, because I haven't been able to fully test it; read on.)  It seems to work, but I can't generate synth audio in CircuitPython fast enough to keep the buffer filled.  That makes the audio generated sound _really_ bad.  Access to the second core might solve this problem.  (There are two things I haven't had time to try yet.  I might be able to get better results using Numpy (whatever it is called in CircuitPython), since it does have trig function generators that might be fast enough, since they are probably implemented in C.  I might also be able to write a whole synth engine module in C (or port a synth virtual processor I wrote in C years ago), but that would be quite a lot more work.  Ideally, I could generate the synth audio on the second core and handle user I/O on the first.  (Note that I haven't even started writing the UI code yet, so even Numpy works in testing, there's no guarantee it could keep up when Circuit Python is also feeding a display and handling input from a bunch of touch buttons.)

**Comment by chrisalbertson** [comment]
Whenever someone says Python is slow compared to C, I ask to look at their code.  Usually, I find they are writing in Python as if in C.

For example, this is “dead dog slow” in Python but fast in C:

For i in range(something):
  Someting(i) = something(i) +1

You should ALWAYS write this as 

Something += 1

If “something” is a numpy array object, the above will likely be as fast as possible on your hardware, possibly even using vectorized math instructions.  The first version does the pointer increment, test, and indexing in the interpreter.  The second version might move all this to hardware if the chip supports it.

Replacing a loop running in an interpreter with a single machine instruction (vector add) might be a 100X speed up.


The above will give you at least 10X or better speed improvement while using two CPU cores will at best only give you a 2X improvement.

Also if you are writing in MicroPython there is one more thing you can do to make it faster.  Use the ***@***.***” code emitter decoration.   This will create optimized native code just like a C++ compiler would generate, so you do use the interpreter at all.

I see zero reason to use Circuitpython.   It is a fork of the main “micropython” that is years behind the current micropython release and is missing a ton of features and optimizations.

You can’t access the second core in circuitpython.  This is one of its missing features.   But multi-tasking in micropython is easy and it works even better than it does under Linux because micropython has no global interpret lock (“GIL”) and allows true multitasking.  One way to do this is with python's ASYNC IO.  so you can do computation and IO at the same time.   

I have a servo controller that is written in 100% micropython that uses numpy and it does the real-time work in one core and the user communication in the other core.  It works well.  The same code runs on ESP32 and RP2020, except the pin numbers are different.   No need to use C.   You can’t do this in circuit python

**Comment by Rybec** [comment]
I am _extremely_ experienced in both Python and C.  I do know how to maximize efficiency in both.  That said, I don't typically use Numpy, because in my day-to-day work, if I need efficiency badly enough to bother with Numpy, I'm better off just using C directly.  Unfortunately, writing C modules for CircuitPython is far more difficult and complicated than writing C modules for desktop Python.  (And half the time, for my job, instead of writing a Python module, I just make a C executable and call it from Python, because even C desktop modules for Python are a pain to write the Python interface code for.  That's not an option in CircuitPython...unless there's something I don't know...)

As far as you seeing "zero reason" to use Circuit, sounds like a person problem.  CircuitPython has better support and a better community for Adafruit products, which are the products I am using.  Just because you can't see the value doesn't mean the value doesn't exist.  (I do have an ESP32 that I might put Micropython on, but that's not a project I have time for now.)

For embedded devices that need high performance, async io is actually _not_ very good when you are running close to the capacity of the device, because it's overhead is too high, and it doesn't have good priority control.  That's why I wrote PyRTOS.  I'm sure async io is significantly more efficient than PyRTOS, which is written in pure Python, but when you need better control over scheduling than async io provides, it's efficiency won't solve your bottleneck.  async io is awesome when you have significantly more CPU power than you need, but it's not so good when you are working within narrow margins.  PyRTOS is good when you need fine priority control and a bit more overhead isn't a huge problem.  When overhead _is_ a problem, and you need fine priority control, the only viable option is manual scheduling.  I haven't found many applications where async io is the best option.  (To be fair, _most_ of my applications need better priority control than async io can offer.  I'm sure async io is quite good for a lot of other applications.)

And as far as servo controllers go, that's an extremely different problem from real-time audio generation.  Nice to hear that Micropython works for you in that, but my audio playback module is written for CircuitPython, and I don't have time to port it to Micropython, regardless of how easy you might believe that to be.  Anyhow, real-time audio generation is far more CPU intensive than controlling a servo.  If it was just simple square waves, it wouldn't be a problem, but a good synth needs to be able to generate _and mix_ trig function waves and others that are far more complex and expensive.  I'm honestly not even sure Numpy will be sufficient for my needs, because I really need to be able to apply an LFO (and other modulation) to the sine wave generation, and I think Numpy only supports the generation of fixed frequency waves.  It is entirely possible that the RP2040 just plain isn't fast enough to do a synth with the full range of capabilities.  If Numpy doesn't work, I _am_ willing to try to do it in C (like I said before, I actually wrote a fairly decent audio processor VM in C, years ago, thought not with wave generator modulation, because I was rather green in audio generation at the time and didn't know how important that was), but that will almost certainly take me years, as I don't have a ton of free time, and writing CircuitPython modules in C is really complicated.

But yes, I do understand that there is at least _a potential_ for over 100 times speed up using Numpy.  That's why I want to try it.  If I can achieve that kind of speed up, then _maybe_ this will work.  It's also possible though, that this isn't the only bottleneck.  If something about reading the Python objects in C is a significant bottleneck, then Numpy will only help and not solve the problem.  Honestly, I probably _will_ get a 100+ times speed up from Numpy.  I don't remember exactly how my code generates the sine wave, but it's probably a list comprehension.  It might be a generator though, because it is critical to preserve state between buffer writes, otherwise it will produce pops in the audio, because the wave segments won't line up.  I'm actually probably using a list comprehension inside of a generator, come to think of it.  While list comprehensions _are_ highly optimized in Python though, they still have significant overhead.  And I don't know how well optimized the trig functions are, but Numpy's trig functions can generate a whole array of outputs at once, rather than one value at a time, which by itself eliminates a ton of function call overhead.  So yeah, I do expect that Numpy will have massively lower overhead, _but_ I'm not sure that will be enough.  (Actually though, I _can_ do modulation with the Numpy sine function.  It takes an array argument, and that means I can populate the array with pre-modulated time/frequency values to achieve modulation.  The last time I had time to work on this was many months ago, so I had forgotten about that.)

Anyhow, no, the problem isn't that I don't know how to write optimal Python code.  Python _is_ slow compared to C.  There are plenty of ways of mitigating this by writing optimal code, but that doesn't magically overcome object and bytecode interpreter overhead.  The problem is that audio synthesis is just very computationally expensive.  I doubt that having a dedicated core would even solve my current problems, but even if I do get my application working by using Numpy, there are still fundamental problems with running the audio generator on the same core, and there are still CPU limitations that could be solved or at least significantly reduced by running the audio generator on its own core.  One of these is complexity.  For example, if I want to apply an LFO to a sine wave, that's a second sine function that has to run, which means I have the overhead of generating two separate sine waves.  If I want to apply effects (amplification, attenuation, clipping...), that's more computation.  If I want to have more than one instrument playing at the same time (to produce chords or harmonies, for example), that's even more computation.  There _will be_ a limit what you can synthesize, on _any_ CPU.  Having a dedicated core for the audio synthesis will significantly raise that limit, allowing the system to synthesize more interesting and complex audio.  My current application is just a fairly simple synth keyboard, but even that needs to be able to generate multiple notes at the same time and mix them (in theory, a user could press all 21 keys at the same time, though I could probably limit playback to only 3 or 4 keys at a time).  It's possible that the RP2040 can't even do that, when I'm also handling keyboard input and running a small I2C display at the same time.  The problem here isn't bad Python code.  It _is_ Python just being slower (and possibly the RP2040 not having the capacity, though I don't _think_ that is the problem).  Numpy might be the solution, or it might not be, but regardless, access to the second RP2040 core would definitely be an improvement, even if it doesn't fully solve the problem.

**Comment by Rybec** [comment]
@chrisalbertson

On a side note, using += is only marginally faster on my machine.  There's a way to go _far_ faster though.

```
for i in range(100000000):
    a = a + 1
```
That takes 7.719s

---

```
for i in range(100000000):
    a += 1
```
This takes 6.479s

---

```
a += sum({1 for i in range(100000000)})
```
That looks more complicated, because it generating a full list and then taking the sum, but it only takes 2.374s

---

If you do a = a + ... on that last one, it takes barely over 3 seconds, but using the list comprehension instead of the for loop still takes less than half the time, even when you don't use +=.  You are totally focusing on the wrong thing for improving efficiency.  Your for loop _is_ the C way of doing it.  The list comprehension is the Python way.

**Comment by chrisalbertson** [comment]
Use the viper code emitter and you can the same back end native code that C uses,

Can you REALLY write C code that is as good as Numpy?   Maybe.  But a lot of smart people have spend time making numpy very fast.   You could duplicate this work if you know about both numeric algorithms and the hardware in the chip.

OK, you had to write an RTOS,  But Micropython already runs on Mbed.  So it has a pretty good RTOS foundation. Multitasking and all.

**Comment by Rybec** [comment]
Well, let's see, I taught several semesters of college level ARM assembly, which included the use of vector instructions.  You _have to be_ good at numeric algorithms to write good audio synthesis code.  The RP2040 has some hardware I'm not that great with (the PIOs mainly), but I don't need to use those for this.

Yeah, I think I could manage.  In fact, I could probably one-up Numpy if I wanted to.  See, Numpy is highly generalized, so that it will work with a broad range of applications.  That almost _always_ comes with performance compromises.  I know exactly what my application is, so I could easily optimize it even further for my specific application.  Honestly, now I really _want_ to just write it all in C and ARM assembly myself, because I _really_ enjoy assembly programming, and I don't have the opportunity to do it anywhere near as often as I would like.


I'm sure Micropython's multitasking is good for most applications.  That's not the problem.  If I'm _writing_ a real-time application, I _need_ significant control over scheduling.  (That's what makes an RTOS and RTOS.)  It doesn't _matter_ if the foundation is RTOS; if I don't have good control over scheduling, it's actually more like a _desktop_ style OS, and it's _not_ an RTOS.  So while I appreciate that Micropython may run within an RTOS, if it doesn't provide RTOS level scheduling control, _it isn't an RTOS itself._  And that means that applications that _need_ an RTOS will _have_ to have an RTOS layer between the OS and the application (which is why I wrote an RTOS for CircuitPython).


Also, adding a second core can provide _far more_ than 2X performance improvement.  You've forgotten that the microcontroller will be running more than just the one task.  If the UI is taking up 90% of the processing power on the CPU, then the audio synthesis is only getting 10%.  Moving it to different core that has nothing running on it gives a _10X_ improvement not a 2X improvement.  This is one of the biggest problems programmers have in understanding efficiency and performance.  _Your application isn't the only thing running on the machine._  Sure, this doesn't apply directly in the case of embedded systems, but it's exactly the same mistake: The task in question isn't the only task the device will be running.  Depending on how much resources are being used by the other tasks, the improvement of moving one task to another core (for that task) could be anywhere between doubling performance and infinity (if it wasn't getting any CPU time at all).  Sure, for my test, where I'm not polling buttons or driving a display, the best I could get from a second core is 2X improvement (though my synth code isn't multi-threaded, so I wouldn't get any improvement _for the test code_), but a synth that plays a single sine wave or that loops through a set of wave forms isn't very useful.  The end product _will_ have a display, 21 keys, and a few more buttons for managing settings.  If the RP2040 uses 50% of a core for that, I'll get a 4X improvement, which is _huge_ in terms of the additional capacity it would give the synth.  That could be 3 to 4 times as many instruments (keys/notes, in this case) playing at the same time.  It could be additional LFOs, to get a more nuanced timbre.  It could allow more effects to be used.  It could open up more effect options, for effects that would be completely unusable sharing a core with the UI code.  And even if the UI only used a marginal amount of CPU time, 2X is still a significant improvement for this application.  (I suspect the UI code will use between 15% and 25%, which is only a small improvement, but it's still enough to justify putting the audio synthesis code in its own core.)  Adding a core is not as simple as "It will only double performance".  You _have to_ consider the wider context to determine the actual effect.  (In this case, it might even actually improve performance on a single task, if Python is doing all of its own management on the first core and the second one is fully free for the task.  I don't know enough about CircuitPython's background management stuff to know if this is the case here, but it very well could be, and that could be a significant performance improvement _just_ by moving the task to the second core that it isn't sharing one with the "OS".)

**Comment by ladyada** [comment]
hi folks, if ya need to chat more please use discord so we can keep this thread on topic.
also, please check out `synthio`, you may be able to add whatever filters/generators to it - we are always happy to look at and help with PRs!

https://github.com/todbot/circuitpython-synthio-tricks

**Comment by Rybec** [comment]
Appologies.  I was not trying to clutter up this issue, and there are some bits that are still relevant.  I'll quiet down in here.

I've looked at synthio before, and it was too primitive for my needs.  Looking at that though, it seems to have come a long way, so I'll give it another look.  If I do go the route of working in C, adding to an existing module would be far simpler for me than trying to make yet another one from scratch.  Thanks for the suggestion!  (Is there somewhere appropriate to discuss synthio in Discord?  I'm probably going to have some questions...)

**Comment by dhalbert** [comment]
Discord is https://adafru.it/discord. The #help-with-circuitpython channel is for help; #circuitpython-dev is appropriate if you want to build CircuitPython or add features.

**Comment by mobigroup** [comment]
@Rybec 
> Python is slow compared to C. 

The statement is obviously false, especially when we discuss Numpy  functions coded in C and Fortran. Moreover, your own C code will always be slower than Numpy code if you do not utilize vectorization and other techniques implemented in Numpy. And with the Numba Python library, we can achieve pure assembler performance (though I suspect it may not be available in Micropython yet). I can certainly back up my claims: you can find my satellite interferometry processor, PyGMTSAR (Python InSAR), on GitHub, which is a much faster alternative to the C-coded GMTSAR. If you can't write effective Python code and need to use C, that's not a problem with Python, and it's off-topic here.

@chrisalbertson 

> Can you REALLY write C code that is as good as Numpy?

Even if we could write C code that matches the performance of Numpy, a minor adjustment to the Numpy code could make it faster again. By the way, do you happen to know if Numba is available in MicroPython or CircuitPython?

**Comment by eightycc** [comment]
Something to consider about module `_thread` in  CPython and MicroPython is that they do not offer concurrent execution of Python code on multiple cores. The interpreter implementation is such that it can run only on one thread at time under the control of the GIL (Global Interpreter Lock). Quote from the CPython documentation:

> CPython implementation detail: In CPython, due to the [Global Interpreter Lock](https://docs.python.org/3/glossary.html#term-global-interpreter-lock), only one thread can execute Python code at once (even though certain performance-oriented libraries might overcome this limitation). If you want your application to make better use of the computational resources of multi-core machines, you are advised to use [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing) or [concurrent.futures.ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor). However, threading is still an appropriate model if you want to run multiple I/O-bound tasks simultaneously.

**Comment by chrisalbertson** [comment]
Circuit Python is Adafruit’s fork of Micro Python.   Micro Python does in fact allow concurrent execution of both cores.

Cicuit Python’s goal is ease of use and providing library functions for all of adafruit’s hardware.   But if your goal is performance, Micro Path is dramatically faster and not just because it can use multiple cores.

That said, many projects don’t need “performance” and ease of use is more important.

I think (?) removing the GIL is on CPyton’s roadmap.  But Python, Circuit Python and micro python are three separate projects and all have different details.

**Comment by chrisalbertson** [comment]
No need for benchmarks just try this in Circuit Pythonand see if it works

@micropython.native or,
@micropython.viper or,
try to use two cores at the same time

Yes they are roughly comparable in performance only as long as you do not use the features present only n Micro Python

As said, each has a different target audience

**Comment by eightycc** [comment]
While it is true that the MP port of RP2 doesn't implement the GIL, I wouldn't be so quick to see this as an advantage. Except for the use case where the spawned thread (the port supports just one of them) can guarantee that it isn't going to touch anything directly or indirectly that will cause both processors to enter multiprocessor unsafe code in core, bad things will happen. Look at the MP ESP32 implementation for a safer and more complete implementation of threads. Yes, it uses the GIL.

**Comment by kamocat** [comment]
I realize there has been a lot of discussion on this, but I wanted to throw in my two cents anyway.

I would be happy enough with programming the second core in C, or even having a fixed set of computation-intensive tasks (for example, FFTs, tensors, cryptographic functions, image processing).
The trouble with running circuitpython on both cores is that it only doubles the computation power at best. I don't need python running on **two** cores if it's unsatisfactory on one. When I'm writing in python, I don't expect it to be "turtles all the way down", but I want the flexibility of python built on top of performant libraries.
In fact, my favorite feature of the rp2040 isn't the second core but the PIO, which gives me low-latency bitbanging without bothering the main processor. And the fact that I can change it on the fly? So cool.
It seems that the fork between Micropython and Circuitpython has diverged so far that merging is difficult and seldom done. That's ok. Maybe the contributors to MicroPython made the wrong decision when including threading and multiprocessing. Maybe you *don't* want to go that route.

**Comment by dhalbert** [comment]
We merge regularly from MicroPython upstream, and plan to PR some internal changes upstream (which are mostly conveniences for the developers).

We have turned off threading and visible support for interrupts in Python, for reasons given here: <https://learn.adafruit.com/cooperative-multitasking-in-circuitpython-with-asyncio#faq-3106700>. However, the support is still latently there.

**Comment by dhalbert** [comment]
> Why didn't you install a SMP RTOS below CircuitPython?

Easier said than done. But we are working in the long-term on zephyr support as a base for most ports.

**Comment by bablokb** [comment]
Zephyr currently also does not support the second core: "As with the Pico 1, there’s no support for running any code on the second core." (from: https://docs.zephyrproject.org/latest/boards/raspberrypi/rpi_pico2/doc/index.html)
